### ISSUE TITLE
async AwsCredentialsProvider?

### ISSUE THREAD

**Issue by dsilvasc** [aws/aws-sdk-java-v2#750]
For cross-cloud workloads, generating AWS credentials can entail [a few different network calls](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_oidc.html).

The interface `AwsCredentialsProvider` provides a single method (`resolveCredentials`) that returns an `AwsCredentials` synchronously.  Although implementations will likely cache credentials, the initial lookup will still block here.

Would it be possible to provide an async AwsCredentialsProvider?

**Comment by millems** [comment]
When you say async, do you mean non-blocking?

**Comment by dsilvasc** [comment]
Yes; I mean an interface that either takes a `Consumer<AwsCredentials>` callback or returns a `CompletionStage<AwsCredentials>`.

**Comment by shorea** [comment]
Due to how our threading model works, I'm not sure if this will be possible. Is it reasonable to do the caching as you mentioned and prime the provider by making the initial call after it's constructed?

FYI for the caching we have this handy util that makes it easier. It's marked as a "protected" API right now but I have no issues making it public (you can even submit a PR if you are participating in Hactoberfest ;)).

https://github.com/aws/aws-sdk-java-v2/blob/master/utils/src/main/java/software/amazon/awssdk/utils/cache/RefreshResult.java

**Comment by dsilvasc** [comment]
Thanks, @shorea.  We have an existing credentials cache implementation that works well with our Scala codebase but we can look into using that one for aws-sdk-java-v2 credentials.

Priming works well for most of our use cases.  We have some others where cold starts are more frequent and priming is thus less effective.
